src/types/type_param.rs

```diff
@@ -62,72 +114,103 @@ impl TypeArg {
         match self {
             TypeArg::Type(s) => Some(s.tag()),
             TypeArg::ClassicType(c) => Some(c.tag()),
+            TypeArg::HashableType(h) => Some(h.tag()),
             _ => None,
         }
     }
 }
 
-/// Checks a [TypeArg] is as expected for a [TypeParam]
-pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgError> {
-    match (arg, param) {
-        (TypeArg::Type(_), TypeParam::Type) => Ok(()),
-        (TypeArg::ClassicType(_), TypeParam::ClassicType) => Ok(()),
-        (TypeArg::HashableType(_), TypeParam::HashableType) => Ok(()),
-        (TypeArg::List(items), TypeParam::List(ty)) => {
-            for item in items {
-                check_type_arg(item, ty.as_ref())?;
+/// A serialized representation of a value of a [CustomType]
+/// restricted to Hashable types.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct CustomTypeArg {
+    /// The type of the constant.
+    /// (Exact matches only - the constant is exactly this type.)
+    typ: CustomType,
+    /// Serialized representation.
+    pub value: serde_yaml::Value,
+}
+
+impl CustomTypeArg {
+    /// Create a new CustomTypeArg. Enforces that the type must be Hashable.
+    pub fn new(typ: CustomType, value: serde_yaml::Value) -> Result<Self, &'static str> {
+        if typ.tag() == TypeTag::Hashable {
+            Ok(Self { typ, value })
+        } else {
+            Err("Only Hashable CustomTypes can be used as TypeArgs")
+        }
+    }
+}
+
+impl ValueOfType for TypeArg {
+    type T = TypeParam;
+
+    fn check_type(&self, ty: &TypeParam) -> Result<(), ConstTypeError> {
+        match (self, ty) {
+            (TypeArg::Type(_), TypeParam::Type) => Ok(()),
+            (TypeArg::ClassicType(_), TypeParam::ClassicType) => Ok(()),
+            (TypeArg::HashableType(_), TypeParam::HashableType) => Ok(()),
+            (TypeArg::Container(vals), TypeParam::Container(c_ty)) => vals.check_container(c_ty),
+            // We might have an argument *value* that is a TypeArg (but not a HashableValue)
```

Comment:
This is the same case as https://github.com/CQCL-DEV/hugr/blob/729e96fc7815f565b5be4d2246ffbb2a0e00f272/src/ops/constant.rs#L181

Reply:
No, this one's bogus, can't happen, because Opaques must be hashable

Reply:
"fail to deduce" where I linked isn't really right - "be unable to represent a container value as hashable, because the value contains opaque constants, and HashableValue deliberately has no way to represent anything opaque"